### PR TITLE
Drain the node before Pod is terminated

### PR DESCRIPTION
**Pull request by zimnx**
Draining node before SIGTERM gracefully shuts down and allows for finishing inflight requests, and prevents new ones.

/cc tnozicka

Part of bigger PR https://github.com/scylladb/scylla-operator/pull/1614

### CODE REVIEW

File: pkg/controller/scyllacluster/resource.go [hunk 1]
```diff
@@ -552,14 +552,14 @@ func StatefulSetForRack(r scyllav1.RackSpec, c *scyllav1.ScyllaCluster, existing
...
-							// need to replay the commitlog in the next startup.
+							// flush the memtable to disk, finish existing requests and stop listening for connections.
+							// Sleep is required to give chance to Load Balancers to acknowledge Pod going down with their
+							// probes.
 							Lifecycle: &corev1.Lifecycle{
 								PreStop: &corev1.LifecycleHandler{
 									Exec: &corev1.ExecAction{
 										Command: []string{
-											"/bin/sh", "-c", "PID=$(pgrep -x scylla);supervisorctl stop scylla; while kill -0 $PID; do sleep 1; done;",
+											"/bin/sh", "-c", "nodetool drain; sleep 15",
```
Comment (tnozicka) [inline pkg/controller/scyllacluster/resource.go:562]: I think any sleep here needs to be configurable, given how LB are (not) integrated / the unpredictability of any network. The approach here also depends on what happens when a new connections arrives after the drain. Does it need to be drained again? Does scylla refuse a new connection after drain? Depending on the answer I see 2 algorithms:

a)
1. `sleep <minTerminationPeriod>`
1. `nodetool drain`

b) 
1. `sleep <minTerminationPeriod> &`
1. `nodetool drain &`
1. `wait`

File: pkg/controller/scyllacluster/resource.go [hunk 2]
```diff
@@ -552,14 +552,14 @@ func StatefulSetForRack(r scyllav1.RackSpec, c *scyllav1.ScyllaCluster, existing
...
-							// This is necessary to ensure we don't lose any data and we don't
-							// need to replay the commitlog in the next startup.
+							// flush the memtable to disk, finish existing requests and stop listening for connections.
+							// Sleep is required to give chance to Load Balancers to acknowledge Pod going down with their
+							// probes.
 							Lifecycle: &corev1.Lifecycle{
 								PreStop: &corev1.LifecycleHandler{
 									Exec: &corev1.ExecAction{
 										Command: []string{
-											"/bin/sh", "-c", "PID=$(pgrep -x scylla);supervisorctl stop scylla; while kill -0 $PID; do sleep 1; done;",
```
Comment (tnozicka) [inline pkg/controller/scyllacluster/resource.go:562]: Can you clarify why this was needed before and not after? Does the SIGTERM (after preStop hook) propagate correctly now?

**Review by tnozicka — CHANGES_REQUESTED** [review]

File: pkg/controller/scyllacluster/resource.go (hunk 2 above)
Comment (tnozicka) [reply pkg/controller/scyllacluster/resource.go:562]: synced on a call, this is fine as SIGTERM now works

File: pkg/controller/scyllacluster/resource.go (hunk 1 above)
Comment (tnozicka) [reply pkg/controller/scyllacluster/resource.go:562]: synced on call, going with b)

File: pkg/controller/scyllacluster/resource.go [hunk 3]
```diff
@@ -552,14 +569,15 @@ func StatefulSetForRack(r scyllav1.RackSpec, c *scyllav1.ScyllaCluster, existing
...
+											"/bin/sh", "-c",
```
Comment (tnozicka) [inline pkg/controller/scyllacluster/resource.go:579]: -eUeo pipefail and err_exit?

File: pkg/api/scylla/v1/types_cluster.go [hunk 4]
```diff
@@ -109,6 +109,12 @@ type ScyllaClusterSpec struct {
 
 	// externalSeeds specifies the external seeds to propagate to ScyllaDB binary on startup as "seeds" parameter of seed-provider.
 	ExternalSeeds []string `json:"externalSeeds,omitempty"`
+
+	// minNodeTerminationPeriod specifies minimum period of time to wait before every drained node is terminated.
+	// This gives time to potential load balancer in front of a node to notice that node is not ready anymore and stop forwarding traffic.
+	// If not provided, Operator will choose this value.
+	// +optional
+	MinNodeTerminationPeriod *metav1.Duration `json:"minNodeTerminationPeriod,omitempty"`
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/api/scylla/v1/types_cluster.go]: done

[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/api/scylla/v1/types_cluster.go]: done

File: pkg/controller/scyllacluster/resource.go (hunk 3 above)
Comment (zimnx) [reply pkg/controller/scyllacluster/resource.go]: done

File: pkg/api/scylla/v1/types_cluster.go [hunk 5]
```diff
@@ -109,6 +109,15 @@ type ScyllaClusterSpec struct {
 
 	// externalSeeds specifies the external seeds to propagate to ScyllaDB binary on startup as "seeds" parameter of seed-provider.
 	ExternalSeeds []string `json:"externalSeeds,omitempty"`
+
+	// minTerminationGracePeriodSeconds specifies minimum duration in seconds to wait before every drained node is
+	// terminated. This gives time to potential load balancer in front of a node to notice that node is not ready anymore
+	// and stop forwarding new requests.
+	// This applies only when node is terminated gracefully.
+	// If not provided, Operator will choose this value.
```
[the remark this answers is not in the document]
Comment (zimnx) [reply pkg/api/scylla/v1/types_cluster.go]: done

**Review by tnozicka — APPROVED** [review]
/lgtm
thanks!

**Comment by zimnx** [comment]
https://github.com/scylladb/scylla-operator/issues/1694#issuecomment-1918746437
https://github.com/scylladb/scylla-operator/issues/1028#issuecomment-1918747044
/retest

**Pull request merged**